feat: rv authentication and extra e2ee

gkc opened the pull request:
Closes #626
Closes #519 

**- What I did**
- Added ability to authenticate to the socket rendezvous, and made this the 
  default behaviour.
- Added ability to end-to-end encrypt all traffic via the socket rendezvous (
  SR), and made this the default behaviour. This provides a good general defense
  against compromise by man-in-the-middle attacks if the two ends are
  communicating via the SR. The encryption is implemented by exchanging a 
  symmetric key via an ephemeral encryption keypair generated by the client 
  for every session.
- Added ability to "ping" a daemon for info about it, including which 
  features the daemon supports.
- By default, sshnp now immediately drops into a prompt for clients that don't
  support sshnp.canRunShell(). `-x` flag allows the ssh command to be output 
  instead of executing ssh immediately.
- Renamed everything sshrvd and sshrv to srvd and srv respectively

**- How to verify it**
- Tests pass

gkc committed:
feat: (idea spike) illustrate how we might use the SocketAuthenticator idea in https://github.com/gkc/socket_connector/tree/socket-authenticator-option

gkc committed:
fix: made code compile
docs: Added TODOs

gkc committed:
docs: added TODOs for sshrvd_impl

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

VJag committed:
Introducing signature verifying authenticator

VJag committed:
Merge remote-tracking branch 'origin/socket-authenticator-option' into socket-authenticator-option

VJag committed:
Refactored code to make it test friendly and added few unit tests

VJag committed:
Refactored code to make it test friendly and added few unit tests

VJag committed:
Changes to send notification as a json

VJag committed:
New abstractions to enable various notification message formats and authentication by SSHRV

VJag committed:
Auth provider and corresponding verification test

VJag committed:
Auth provider and corresponding verification test

VJag committed:
Authenticate in Sshrv when authenticationProvider is provided

VJag committed:
Changes to refactor notification message classes and introduced manager class to create the right message type

VJag committed:
Changes to refactor notification message classes and introduced manager class to create the right message type

VJag committed:
Changed few important class names. The changes in this commit are a result of many a files uptaking the new name.

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

# Conflicts:
#	packages/noports_core/lib/src/sshnp/impl/sshnp_openssh_local_impl.dart
#	packages/noports_core/lib/src/sshrvd/sshrvd_impl.dart

gkc committed:
chore: fixed lint warnings

VJag committed:
Minor changes to improve the testability

VJag committed:
Merge remote-tracking branch 'origin/socket-authenticator-option' into socket-authenticator-option

VJag committed:
This commit contains:

1. Changes to take args to accept authenticateClient and authenticateDevice args
2. Using the values of these args to make decisions
3. Changes to the way we sign the data

VJag committed:
made a null aware invocation

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

gkc committed:
feat: rvd authentication working end to end
- new sshrv parameters --bind-local-port and --rvd-auth-string
- uses new socket_connector method, serverToSocket
- multiple other tweaks to get things to work

gkc committed:
interim commit: proves end-to-end working using hard-coded AES key and IV in sshrv. Next commit will have the AES key and IV creation, sharing, and passing in to the sshrv process via command line and parameters

gkc committed:
fix: remove unnecessary delays when sending notifications.
- add checkForFinalDeliveryStatus and waitForFinalDeliveryStatus parameters everywhere we call `notify`
- make callers use those parameters appropriately.
Note that in no case was there a need for those parameters to be true, but it seems better to introduce the parameters and make callers explicitly aware of them, rather than the simpler change to just set them to false in every case.

gkc committed:
fix: remove 9.7s of unnecessary delay from a unit test

gkc committed:
feat: Client ability to ping daemon for info, including which features the daemon supports. This is then used to downgrade where the daemon does not support the feature

gkc committed:
fix: something weird about error handling in the sshnp client; commented out two 'return' statements for now, in the runZonedGuarded onError handler

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

gkc committed:
feat: add `--discover-daemon-features` to sshnp args

gkc committed:
chore: remove ephemeralEncryptionKeyPair from daemon; client will generate it instead, so that a ping to the daemon isn't required in order to share session encryption keys

gkc committed:
feat: add encryptRvdTraffic flag and ephemeral public key to ssh session request message from client.

gkc committed:
feat: sshrv use AES key and IV, when supplied, for en/decryption

gkc committed:
fix: Need to give the sshrv process time to start and bind to its port before starting the initial tunnel which uses that port

gkc committed:
fix: Add defensive code to SshnpDartPureImpl to un-set the new flags, as SshnpDartPureImpl does not yet support them.

gkc committed:
build: update pubspec to use git branch for socket_connector instead of local relative path

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

# Conflicts:
#	packages/noports_core/lib/src/version.dart
#	packages/noports_core/pubspec.yaml
#	packages/sshnoports/pubspec.lock
#	packages/sshnoports/pubspec.yaml

gkc committed:
feat: encrypt/decrypt aes key and IV for rvd traffic using session ephemeral keypair

gkc committed:
refactor: streamified socket_connector

gkc committed:
feat: limited info leakage by moving sshrv params from command line to env vars

gkc committed:
chore: pinned some dependencies
feat: increased timeout for responses from sshnpd and sshrvd

gkc committed:
chore: more logging cleanup

gkc committed:
chore: more logging cleanup

gkc commented on packages/noports_core/lib/src/common/at_ssh_key_util/local_ssh_key_util.dart:
```diff
@@ -129,7 +129,8 @@ class LocalSshKeyUtil implements AtSshKeyUtil {
     // supported keys by the dartssh2 package
-    if (!sshPublicKey.startsWith(RegExp(r'^(ecdsa-sha2-nistp)|(rsa-sha2-)|(ssh-rsa)|(ssh-ed25519)|(ecdsa-sha2-nistp)'))) {
+    if (!sshPublicKey.startsWith(RegExp(
+        r'^(ecdsa-sha2-nistp)|(rsa-sha2-)|(ssh-rsa)|(ssh-ed25519)|(ecdsa-sha2-nistp)'))) {
```
this was just dart format

gkc commented on packages/noports_core/lib/src/common/default_args.dart:
```diff
@@ -18,6 +18,10 @@ class DefaultArgs {
+  static const bool authenticateClientToRvd = true;
+  static const bool authenticateDeviceToRvd = true;
+  static const bool encryptRvdTraffic = true;
+  static const bool discoverDaemonFeatures = true;
```
extra command line flags

gkc commented on packages/noports_core/lib/src/common/mixins/at_client_bindings.dart:
```diff
@@ -3,19 +3,26 @@ import 'package:at_utils/at_utils.dart';
+    String value, {
+    required bool checkForFinalDeliveryStatus,
+    required bool waitForFinalDeliveryStatus,
+  }) async {
```
I added these params as required and dealt with the ripple effects as it is better in general to have these parameters explicitly exposed

gkc commented on packages/noports_core/lib/src/common/types.dart:
```diff
@@ -1,12 +1,21 @@
+typedef SshrvGenerator<T> = Sshrv<T> Function(
+  String,
+  int, {
+  required int localPort,
```
renamed localSshdPort to localPort in anticipation of future daemon versions which connect to ports other than the sshd port.

Also made it `required` to make sure all callers are being explicit

gkc commented on packages/noports_core/lib/src/common/types.dart:
```diff
@@ -1,12 +1,21 @@
+  String,
+  int, {
+  required int localPort,
+  required bool bindLocalPort,
```
Client-side we now run an rv which binds to a local port on one side, and talks to the rvd on the other; the ssh tunnel is then created by sshing to the local port that the local sshrv is listening on

gkc commented on packages/noports_core/lib/src/common/types.dart:
```diff
@@ -1,12 +1,21 @@
+  required bool bindLocalPort,
+  String? rvdAuthString,
+  String? sessionAESKeyString,
+  String? sessionIVString,
```
Parameters for auth and e2ee, for connections to rvd

gkc commented on packages/noports_core/lib/src/sshnp/impl/sshnp_dart_pure_impl.dart:
```diff
@@ -2,14 +2,35 @@ import 'dart:async';
+    if (params.authenticateClientToRvd) {
+      logger.shout('$runtimeType: disabling authenticateClientToRvd flag');
+      params.authenticateClientToRvd = false;
+    }
```
As mentioned above - the major work outstanding for these features is to be able to an RV in-memory, which we need to be able to do for pure dart apps e.g. the noports gui etc

xavierchanth replied:
As mentioned in one of my other comments, it would be useful to have an in-memory RV with the openssh client as well, then we can call the ssh command as a child process using Process.run()

gkc commented on packages/noports_core/lib/src/sshnp/impl/sshnp_openssh_local_impl.dart:
```diff
@@ -66,6 +67,10 @@ class SshnpOpensshLocalImpl extends SshnpCore
 
+    final server = await ServerSocket.bind(InternetAddress.anyIPv4, 0);
+    int localRvPort = server.port;
+    await server.close();
```
Above code is to identify a port that is available for the rv process to bind to

gkc commented on packages/noports_core/lib/src/sshnp/impl/sshnp_openssh_local_impl.dart:
```diff
@@ -94,6 +109,16 @@ class SshnpOpensshLocalImpl extends SshnpCore
+      localRvPort: localRvPort,
+      sessionAESKeyString: sshnpdChannel.sessionAESKeyString,
+      sessionIVString: sshnpdChannel.sessionIVString,
+    );
```
We now run an rv on the client side

gkc commented on packages/noports_core/lib/src/sshnp/impl/sshnp_openssh_local_impl.dart:
```diff
@@ -94,6 +109,16 @@ class SshnpOpensshLocalImpl extends SshnpCore
+      sessionIVString: sshnpdChannel.sessionIVString,
+    );
+
+    await Future.delayed(Duration(seconds: 1));
```
Had to add this in order to give the sshrv process time to start up and bind to its port

xavierchanth replied:
Can't we observe the process in which we start up sshrv and wait for that?

xavierchanth replied:
nvm... we would have to make that process fork and return an exit code, which we currently don't do. (like ssh with ForkAfterAuthentication enabled)

gkc commented on packages/noports_core/lib/src/sshnp/impl/sshnp_unsigned_impl.dart:
```diff
@@ -94,6 +95,9 @@ class SshnpUnsignedImpl extends SshnpCore with SshnpLocalSshKeyHandler {
     }
 
+    /// Start sshrv
+    var bean = await sshrvdChannel.runSshrv(directSsh: false);
```
uhmmm this might have been added in error, will need to check it

xavierchanth replied:
Unsigned client always ran sshrv, this is the legacy v3 client. I see that it has been moved to start up after the notification for an sshd request is sent, this could lead to the daemon attempting to ssh connect back to the client before sshrv is open (but in the real world, I don't think this is likely to happen)

gkc commented on packages/noports_core/lib/src/sshnp/models/sshnp_params.dart:
```diff
@@ -319,8 +379,12 @@ class SshnpPartialParams {
+          : AtUtils.fixAtSign(args[SshnpArg.fromArg.name]),
+      sshnpdAtSign: args[SshnpArg.toArg.name] == null
+          ? null
+          : AtUtils.fixAtSign(args[SshnpArg.toArg.name]),
```
Need atSigns to consistently have the `@` prefixed

xavierchanth replied:
We dropped support for non-atSign hosts, so I think we can enforce the check for those too.

gkc commented on packages/noports_core/lib/src/sshnp/util/sshrvd_channel/sshrvd_channel.dart:
```diff
@@ -31,27 +33,34 @@ abstract class SshrvdChannel<T> with AsyncInitialization, AtClientBindings {
 
   String? _host;
-  int? _port;
+  int? _portA;
```
Renamed to help readability

gkc commented on packages/noports_core/lib/src/sshnp/util/sshrvd_channel/sshrvd_channel.dart:
```diff
@@ -31,27 +33,34 @@ abstract class SshrvdChannel<T> with AsyncInitialization, AtClientBindings {
 
   /// The port sshrvd is listening on
-  int? _sshrvdPort;
+  int? _portB;
```
Renamed to help readability

gkc commented on packages/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -329,9 +341,10 @@ class SshnpdImpl implements Sshnpd {
+      // Check to see if the ssh public key is
+      // supported keys by the dartssh2 package
+      if (!sshPublicKey.startsWith(RegExp(
+          r'^(ecdsa-sha2-nistp)|(rsa-sha2-)|(ssh-rsa)|(ssh-ed25519)|(ecdsa-sha2-nistp)'))) {
```
no change, just dart format

gkc commented on packages/noports_core/lib/src/sshrv/sshrv_impl.dart:
```diff
@@ -34,10 +58,36 @@ class SshrvImplExec implements Sshrv<Process> {
+    Map<String, String> environment = {};
+    if (rvdAuthString != null) {
+      rvArgs.addAll(['--rv-auth']);
+      environment['RV_AUTH'] = rvdAuthString!;
```
Adding to environment rather than command line to prevent any potentially interesting info from being unnecessarily visible

xavierchanth replied:
We should have good documentation explaining how the flag + environment variable works in tandem

xavierchanth replied:
I think including this in the CLI printed usage is good enough, but it needs to be well written.

gkc commented on packages/noports_core/lib/src/sshrv/sshrv_impl.dart:
```diff
@@ -34,10 +58,36 @@ class SshrvImplExec implements Sshrv<Process> {
+    if (sessionAESKeyString != null && sessionIVString != null) {
+      rvArgs.addAll(['--rv-e2ee']);
+      environment['RV_AES'] = sessionAESKeyString!;
+      environment['RV_IV'] = sessionIVString!;
```
Adding to environment rather than command line to prevent any potentially interesting info from being unnecessarily visible

gkc commented on packages/noports_core/lib/src/sshrvd/build_env.dart:
```diff
@@ -1,4 +1,5 @@
 class BuildEnv {
-  static final bool enableSnoop =
-      bool.fromEnvironment('ENABLE_SNOOP', defaultValue: false);
+  static final bool enableSnoop = (Platform.environment['ENABLE_SNOOP'] ?? "false").toLowerCase() == 'true';
```
I found that `bool.fromEnvironment` was not working properly on MacOS

xavierchanth replied:
We should probably file a bug with the dart team

gkc commented on packages/noports_core/lib/src/sshrvd/signature_verifying_socket_authenticator.dart:
```diff
@@ -0,0 +1,118 @@
+        sc.add(data);
+      } else {
+        try {
+          final message = String.fromCharCodes(data);
```
TODO: This code currently assumes that all of the authentication challenge response comes in a single flow. Ideally should parse the ints until we get a newline, then
- use everything up to the newline for auth
- discard the newline
- add everything after the newline to the output stream via `sc.add`

gkc commented on packages/noports_core/lib/src/sshrvd/socket_connector.dart:
```diff
@@ -1,48 +1,123 @@
-    serverPortB: portB,
-    verbose: snoop,
-  );
+  SocketConnector connector = await SocketConnector.serverToServer(
```
Renamed here (and other places) to 'connector'; the name `socketStream` was confusing me since SocketConnector is not a stream of Sockets

gkc commented on packages/noports_core/lib/src/sshrvd/socket_connector.dart:
```diff
@@ -1,48 +1,123 @@
   while (closed == false) {
-    closed = await socketStream.closed();
+    logger.info('Waiting for connector to close');
+    closed = await connector.closed();
```
TODO don't need a `while` loop here

gkc commented on packages/noports_core/pubspec.yaml:
```diff
@@ -24,6 +24,12 @@ dependencies:
   socket_connector: ^1.0.11
```
Set to 1.1.0 once socket_connector PR has been merged

gkc commented on packages/noports_core/pubspec.yaml:
```diff
@@ -24,6 +24,12 @@ dependencies:
+  socket_connector:
+    git:
+      url: https://github.com/gkc/socket_connector.git
+      ref: socket-authenticator-option
```
Remove once socket_connector PR has been merged and the package has been published

gkc committed:
feat: supply verbose and logTraffic parameters separately to `SocketConnector.serverToServer`

gkc committed:
refactor: changed variable names

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

xavierchanth commented on packages/noports_core/lib/src/sshnp/util/sshnpd_channel/sshnpd_channel.dart:
```diff
@@ -87,16 +87,16 @@ abstract class SshnpdChannel with AsyncInitialization, AtClientBindings {
   /// Returns false if a timeout occurred.
-  Future<SshnpdAck> waitForDaemonResponse() async {
+  Future<SshnpdAck> waitForDaemonResponse({int maxWaitMillis = 15000}) async {
     // Timer to timeout after 10 Secs or after the Ack of connected/Errors
```
This comment is out of sync with the changes, maybe change it to say
```dart
/// Timer to timeout after [maxWaitMillis] milliseconds or after the Ack of connected/Errors
```

gkc replied:
Done

xavierchanth commented on packages/noports_core/lib/src/sshrvd/signature_verifying_socket_authenticator.dart:
```diff
@@ -40,60 +42,75 @@ class SignatureAuthVerifier implements SocketAuthVerifier {
+          // "hashingAlgo":"<algo>",
+          // "signingAlgo":"<algo>",
+          // "payload":{<the data which was signed>}
+          // }
```
We should dartdoc (`///`) this comment block with message

gkc replied:
Done

xavierchanth commented on packages/noports_core/lib/src/sshnp/util/sshnpd_channel/sshnpd_default_channel.dart:
```diff
@@ -66,8 +68,16 @@ mixin SshnpdDefaultPayloadHandler on SshnpdChannel {
+
+      sessionIVString = daemonResponse['sessionIV'];
+      logger.info('Received sessionIV: $sessionIVString');
+
```
Should we be logging these values?

xavierchanth replied:
Could pose a risk to someone who is screen sharing, or sharing logs for debugging...

gkc replied:
Now logging the encrypted values

xavierchanth commented on packages/noports_core/lib/src/common/features.dart:
```diff
@@ -0,0 +1,8 @@
+
+  /// End-to-end encrypt traffic sent via the SocketRendezvous (sr)
+  srE2ee,
+}
```
Should we add share public key to this list of features? (i.e. whether -s flag is on or off)

gkc replied:
Good idea!

xavierchanth commented on packages/noports_core/lib/src/common/validation_utils.dart:
```diff
@@ -1,4 +1,5 @@
 import 'dart:convert';
+import 'dart:io';
```
I was very meticulous to wrap all io operations into src/common/io_types.dart and src/common/file_system_utils.dart so that we can mock it in the unit tests and ensure that we have virtualized operations for the pure dart client.

Ideally, we don't import dart:io outside of these files I mentioned above.

gkc replied:
was an unnecessary import, fixed

xavierchanth commented on packages/noports_core/lib/src/sshnp/models/sshnp_arg.dart:
```diff
@@ -350,18 +351,28 @@ class SshnpArg {
   );
   static const authenticateClientToRvdArg = SshnpArg(
```
Can we add an abbreviation to these authentication flags?

Maybe -a for client auth, and -A for daemon auth?

gkc replied:
done

xavierchanth commented on packages/noports_core/lib/src/sshrvd/socket_connector.dart:
```diff
@@ -103,10 +106,9 @@ void socketConnector(ConnectorParams connectorParams) async {
+      ' for ${sshrvdSessionParams.atSignA} to ${sshrvdSessionParams.atSignB}'
+      ' using ports [$portA, $portB]');
 
   Isolate.current.kill();
```
Question about isolates as we scale this:

What happens to the memory of the current isolate when we kill it, and how does the GC handle it?

gkc replied:
Maybe something that @cconstab has looked at on the production SR daemons?

xavierchanth commented on packages/noports_core/lib/src/sshnp/impl/notification_request_message.dart:
```diff
@@ -1,33 +1,29 @@
+        'authenticateToRvd': authenticateToRvd,
+        'clientNonce': clientNonce,
+        'rvdNonce': rvdNonce,
+      };
```
Going forward, we might want to consider a lightweight JSON spec for these various payloads - considering we have a Dart & Python impl, with C on the way.

xavierchanth replied:
This might cause more tech debt, without any real benefit though...

gkc replied:
I've created [a ticket](https://github.com/atsign-foundation/noports/issues/692) for now

xavierchanth commented on packages/noports_core/lib/src/sshrvd/build_env.dart:
```diff
@@ -1,4 +1,4 @@
-  static final bool enableSnoop =
-      bool.fromEnvironment('ENABLE_SNOOP', defaultValue: false);
+  static final bool enableSnoop = true;
+      // bool.fromEnvironment('ENABLE_SNOOP', defaultValue: false);
```
Are we eliminating the debug build?

gkc replied:
this was just for my convenience while actively developing this PR's features

xavierchanth commented on packages/noports_core/lib/src/sshrv/sshrv_impl.dart:
```diff
@@ -73,13 +73,18 @@ class SshrvImplDart implements Sshrv<SocketConnector> {
+      // TBD - Gary/Xavier/Jagan
+      // Should we expose the sockets from socketConnector? whats the best way to do this.
+      //authenticationProvider?.authenticate(socketConnector.socketB);
+      return socketConnector;
```
I am mostly unopinionated on this, but we should have some way to kill the socket connector externally, I have thought about some use-cases where this is useful.

gkc replied:
one can call close() on a socket connector

xavierchanth commented on packages/noports_core/lib/src/sshnp/models/sshnp_arg.dart:
```diff
@@ -346,4 +350,44 @@ class SshnpArg {
+    parseWhen: ParseWhen.commandLine,
+    mandatory: false,
+  );
+  static const discoverDaemonFeaturesArg = SshnpArg(
```
would be nice to have an abbreviation for this as well.

gkc replied:
Have added as 'F'
Have also added abbreviation 'E' for 'encrypt-rvd-traffic'

xavierchanth commented:
@gkc @VJag general feedback:

- Make sure to wrap all dart:io calls as I had previously done with io_types.dart and file_system_utils.dart, we need everything wrapped for two reasons:
  - unit testing with mock io calls
  - the pure-dart client must not depend on any io calls, so we must avoid leaking io calls into the base classes
- end2end tests need to be fixed, we currently build in a docker image from a branch and then create run images from that, instead we should download from the releases, extract and call the binary from there. We also need to define testing matrices which understand the differences between version and test against older versions / limited features sets correctly.

xavierchanth committed:
Merge remote-tracking branch 'origin/trunk' into sao-merge-renames

gkc committed:
Merge pull request #669 from atsign-foundation/sao-merge-renames

chore: socket-authenticator-option merge renames

gkc committed:
Merge remote-tracking branch 'origin/socket-authenticator-option' into socket-authenticator-option

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

# Conflicts:
#	packages/dart/noports_core/lib/src/sshnp/impl/sshnp_openssh_local_impl.dart
#	packages/dart/noports_core/lib/src/version.dart
#	packages/dart/noports_core/pubspec.yaml
#	packages/dart/sshnoports/pubspec.lock
#	packages/dart/sshnoports/pubspec.yaml

gkc committed:
feat: incorporated latest socket_connector changes

gkc committed:
style: ran dart format

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

gkc committed:
fix: handle empty string for tunnelUserName parameter as if it had been presented as `null`

gkc committed:
feat: performance improvements
1. have SshrvdImpl fetch the public key of another atSign via the Sshrvd's own atClient
2. instead of waiting 1 second for a forked sshrv process to have started up, instead listen to its stderr and wait for the message that says "rv is running"

gkc committed:
style: ran dart format

gkc committed:
chore: tidy up logging

gkc committed:
chore: adjusted SshrvdUtil test following refactoring of static methods to instance methods

gkc committed:
feat: default new flags authenticateClientToRvd, authenticateDeviceToRvd, encryptRvdTraffic and discoverDaemonFeatures to false

gkc committed:
feat: compatibility with old sshrvd versions
- if new rvd features (authentication) are NOT being used, send an old-style request
- if new rvd features (authentication) ARE being used, send a new-style request

gkc committed:
chore: fix pubspec.lock for melos build in github actions

gkc committed:
chore: trying to find what the problem is with e2e local-local failing because of timeout waiting for sshrvd response

gkc committed:
fix: make new sshrvdChannel handle responses from older versions of sshrvd

gkc committed:
style: ran dart format

gkc committed:
feat: address review comments
- in ping response, daemon says whether or not it accepts public keys from the client
- added abbreviations in SshnpArg for the new flags
- corrected some code comments and added to others

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

xavierchanth committed:
chore: rename sshrv(d) to srv(d) in file names

xavierchanth committed:
refactor: make srv look for old and new binary name on the path

xavierchanth committed:
chore: rename Sshrv and SshrvImpls

xavierchanth committed:
fix: bug prone string matching

xavierchanth committed:
chore: rename the rest of sshrv(d) to srv(d)

xavierchanth committed:
chore: rename the remaining files from sshrv(d) to srv(d)

xavierchanth committed:
chore: rename remaining sshrv files

xavierchanth committed:
feat(BREAKING CHANGE): add -x flag

By default sshnp now immediately drops into a prompt for clients that don't support sshnp.canRunShell(). -x allows the ssh command to be output instead of executing ssh immediately.

xavierchanth committed:
fix: add stdout to ChildProcessAsSshnpRemoteProcess

xavierchanth committed:
fix: use standalone Process.start for ssh child process

xavierchanth committed:
chore: cleanup -x flag

gkc committed:
Merge branch 'socket-authenticator-option' into execution-flag

gkc committed:
Merge pull request #703 from atsign-foundation/rename-sr

refactor(BREAKING CHANGE): Rename sshrv to srv

gkc committed:
Merge remote-tracking branch 'origin/trunk' into socket-authenticator-option

gkc committed:
Merge branch 'socket-authenticator-option' into execution-flag

gkc committed:
Merge pull request #705 from atsign-foundation/execution-flag

feat(BREAKING CHANGE): add -x flag

xavierchanth committed:
ci: multibuild using local noports_core

xavierchanth committed:
ci: update macos arm64 build script to use local noports_core

xavierchanth committed:
ci: update build dependencies

gkc committed:
feat: set new flags to default to true, as we are releasing a new major version

xavierchanth committed:
fix: x flag

gkc committed:
docs: Modified, and added some further explanation of the reason for, SrvImplExec.completionString

xavierchanth committed:
fix: multibuild upload path for non x64 builds

gkc committed:
fix: catch FormatException in SrvdImpl when parsing args so that we can gracefully deal with unknown options (many of which people are likely to try, such as `--version` or `--help`

gkc committed:
feat: sshnpd: Added version to abstract class Sshnpd and various other places; this is so that the daemon can respond with its released version as well as the version of the no_ports_core package
fix: sshnpd: catch FormatException in SshnpdImpl when parsing args so that we can gracefully deal with unknown options (many of which people are likely to try, such as `--version` or `--help`)
fix: sshnpd: ensured that the heartbeat's device info is the same as what is sent in a `ping` response
docs: updated CHANGELOG
fix: added a usage printer to the srv binary
fix: standardized usage printing and version printing, so they always go to stderr

xavierchanth committed:
Merge branch 'socket-authenticator-option' into ci-allow-local-multi-build

gkc committed:
fix: add a little more arg / env checking to the srv programme

gkc committed:
Merge pull request #707 from atsign-foundation/fix-x-flag

fix: x flag

gkc committed:
Merge pull request #711 from atsign-foundation/ci-allow-local-multi-build

ci: allow local multi build

gkc committed:
fix: modifying sshnp_entrypoint.sh so at least the local-local e2e test passes

gkc committed:
Merge remote-tracking branch 'origin/socket-authenticator-option' into socket-authenticator-option

gkc committed:
test: temporarily disabling the end2end_tests.yaml workflow

gkc committed:
feat: include the core package version in usage info of the binaries
